Estimate gas for 'approve' contract call

sergeyboyko0791 opened the pull request:
Estimate gas for 'approve' contract call instead of using a hardcoded 50k gas limit
* Add 'EthCoin::estimate_gas_for_contract_call'

sergeyboyko0791 commented:
@artemii235 I would suggest using estimation gas instead of hardcoded gas limits for all EtomicSwap contract calls.
If you are agree, I'll do it in this PR.

artemii235 commented:
@sergeyboyko0791 Yes, it will be nice to have and it's good opportunity to implement :slightly_smiling_face:

shamardy commented on mm2src/coins/eth.rs:
```diff
@@ -1646,18 +1672,27 @@ impl EthCoin {
-                self.sign_and_send_transaction(0.into(), Action::Call(*token_addr), data, U256::from(APPROVE_GAS_LIMIT))
-            },
-        }
+            let gas_limit = try_s!(
```
We shouldn't use `try_s!` in new code since it's deprecated

sergeyboyko0791 replied:
You are right! Thanks)
Will remove it. Also, I'll ask you to re-review the PR again when I finish the following :)
> I would suggest using estimation gas instead of hardcoded gas limits for all EtomicSwap contract calls

shamardy reviewed (commented):
Thanks for the PR! Only one comment 🙂

artemii235 commented:
@sergeyboyko0791 Is this still a draft? :slightly_smiling_face:

sergeyboyko0791 commented:
> Is this still a draft? slightly_smiling_face

It works only for `approve` contract call. Everything else is in progress yet. I've postponed it due to the last release.

shamardy reviewed (approved):
LGTM! only removal of `try_s!` from the last review is what's left

sergeyboyko0791 commented:
> LGTM! only removal of `try_s!` from the last review is what's left

I forgot about this since I replaced it locally already when integrating the estimation of gas for every ETH contract call. There I added a few new error types.
Could we merge it without these changes? I'll definitely finish it later.

shamardy commented:
Yeah this is can be merged for sure, `try_s!` is a negligible issue and it's still used in other places anyway.

artemii235 commented on mm2src/coins/eth.rs:
```diff
@@ -1646,18 +1672,27 @@ impl EthCoin {
     }
 
     fn approve(&self, spender: Address, amount: U256) -> EthTxFut {
```
I have remembered one thing: could you please call the `approve` with `amount = U256::max`? Most ETH/BNB dApps, that I've tried, use the same approach. It saves some gas by avoiding repetitive approval calls.

artemii235 reviewed (changes requested):
One minor change proposal.
